File: include/webengine/libqwebengineframe.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://doc.qt.io/qt-6/qwebengineframe.html
pub const qwebengineframe = struct {
    /// New constructs a new QWebEngineFrame object.
    ///
    /// ``` param1: QtC.QWebEngineFrame ```
    pub fn New(param1: ?*anyopaque) QtC.QWebEngineFrame {
        return qtc.QWebEngineFrame_new(@ptrCast(param1));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineframe.html#isValid)
    ///
    /// ``` self: QtC.QWebEngineFrame ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return qtc.QWebEngineFrame_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineframe.html#name)
    ///
    /// ``` self: QtC.QWebEngineFrame, allocator: std.mem.Allocator ```
    pub fn Name(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWebEngineFrame_Name(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qwebengineframe.Name: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineframe.html#htmlName)
    ///
    /// ``` self: QtC.QWebEngineFrame, allocator: std.mem.Allocator ```
    pub fn HtmlName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWebEngineFrame_HtmlName(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qwebengineframe.HtmlName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineframe.html#children)
    ///
    /// ``` self: QtC.QWebEngineFrame, allocator: std.mem.Allocator ```
    pub fn Children(self: ?*anyopaque, allocator: std.mem.Allocator) []QtC.QWebEngineFrame {
        const _arr: qtc.libqt_list = qtc.QWebEngineFrame_Children(@ptrCast(self));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.QWebEngineFrame, _arr.len) catch @panic("qwebengineframe.Children: Memory allocation failed");
        const _data: [*]QtC.QWebEngineFrame = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineframe.html#url)
    ///
    /// ``` self: QtC.QWebEngineFrame ```
    pub fn Url(self: ?*anyopaque) QtC.QUrl {
        return qtc.QWebEngineFrame_Url(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineframe.html#size)
    ///
    /// ``` self: QtC.QWebEngineFrame ```
    pub fn Size(self: ?*anyopaque) QtC.QSizeF {
        return qtc.QWebEngineFrame_Size(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineframe.html#isMainFrame)
    ///
    /// ``` self: QtC.QWebEngineFrame ```
    pub fn IsMainFrame(self: ?*anyopaque) bool {
        return qtc.QWebEngineFrame_IsMainFrame(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineframe.html#runJavaScript)
    ///
    /// ``` self: QtC.QWebEngineFrame, script: []const u8 ```
    pub fn RunJavaScript(self: ?*anyopaque, script: []const u8) void {
        const script_str = qtc.libqt_string{
            .len = script.len,
            .data = script.ptr,
        };
        qtc.QWebEngineFrame_RunJavaScript(@ptrCast(self), script_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineframe.html#printToPdf)
    ///
    /// ``` self: QtC.QWebEngineFrame, filePath: []const u8 ```
    pub fn PrintToPdf(self: ?*anyopaque, filePath: []const u8) void {
        const filePath_str = qtc.libqt_string{
            .len = filePath.len,
            .data = filePath.ptr,
        };
        qtc.QWebEngineFrame_PrintToPdf(@ptrCast(self), filePath_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineframe.html#runJavaScript)
    ///
    /// ``` self: QtC.QWebEngineFrame, script: []const u8, worldId: u32 ```
    pub fn RunJavaScript22(self: ?*anyopaque, script: []const u8, worldId: u32) void {
        const script_str = qtc.libqt_string{
            .len = script.len,
            .data = script.ptr,
        };
        qtc.QWebEngineFrame_RunJavaScript22(@ptrCast(self), script_str, @intCast(worldId));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineframe.html#dtor.QWebEngineFrame)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QWebEngineFrame ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QWebEngineFrame_Delete(@ptrCast(self));
    }
};

File: src_original/webengine/libqwebengineframe.cpp
#include <QList>
#include <QSizeF>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QUrl>
#include <QWebEngineFrame>
#include <qwebengineframe.h>
#include "libqwebengineframe.h"
#include "libqwebengineframe.hxx"

QWebEngineFrame* QWebEngineFrame_new(const QWebEngineFrame* param1) {
    return new QWebEngineFrame(*param1);
}

bool QWebEngineFrame_IsValid(const QWebEngineFrame* self) {
    return self->isValid();
}

libqt_string QWebEngineFrame_Name(const QWebEngineFrame* self) {
    QString _ret = self->name();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string QWebEngineFrame_HtmlName(const QWebEngineFrame* self) {
    QString _ret = self->htmlName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_list /* of QWebEngineFrame* */ QWebEngineFrame_Children(const QWebEngineFrame* self) {
    QList<QWebEngineFrame> _ret = self->children();
    // Convert QList<> from C++ memory to manually-managed C memory
    QWebEngineFrame** _arr = static_cast<QWebEngineFrame**>(malloc(sizeof(QWebEngineFrame*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new QWebEngineFrame(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

QUrl* QWebEngineFrame_Url(const QWebEngineFrame* self) {
    return new QUrl(self->url());
}

QSizeF* QWebEngineFrame_Size(const QWebEngineFrame* self) {
    return new QSizeF(self->size());
}

bool QWebEngineFrame_IsMainFrame(const QWebEngineFrame* self) {
    return self->isMainFrame();
}

void QWebEngineFrame_RunJavaScript(QWebEngineFrame* self, const libqt_string script) {
    QString script_QString = QString::fromUtf8(script.data, script.len);
    self->runJavaScript(script_QString);
}

void QWebEngineFrame_PrintToPdf(QWebEngineFrame* self, const libqt_string filePath) {
    QString filePath_QString = QString::fromUtf8(filePath.data, filePath.len);
    self->printToPdf(filePath_QString);
}

void QWebEngineFrame_RunJavaScript22(QWebEngineFrame* self, const libqt_string script, unsigned int worldId) {
    QString script_QString = QString::fromUtf8(script.data, script.len);
    self->runJavaScript(script_QString, static_cast<quint32>(worldId));
}

void QWebEngineFrame_Delete(QWebEngineFrame* self) {
    delete self;
}
